integration-tests/oidc-code-flow/src/main/java/io/quarkus/it/keycloak/MultipleAuthMechanismResource.java

```diff
@@ -0,0 +1,29 @@
+package io.quarkus.it.keycloak;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+import io.quarkus.oidc.CodeFlow;
+import io.quarkus.security.Authenticated;
+import io.quarkus.security.PermissionsAllowed;
+import io.quarkus.vertx.http.runtime.security.Basic;
+
+@Path("multiple-auth-mech")
+public class MultipleAuthMechanismResource {
+
+    @Basic
+    @PermissionsAllowed("permission1")
+    @GET
+    @Path("basic")
+    public String basicAuthMech() {
+        return "basicAuthMech";
+    }
+
+    @CodeFlow
+    @Authenticated
```

Comment:
Would be great for `@Authenticated` be assumed when `@Basic` or `@CodeFlow` is used

Reply:
As far as `Bearer`/`CodeFlow` is concerned, how would it work in a multi-tenant OIDC setup ?

Reply:
Please keep in mind that you can select authentication mechanism used when `SecurityIdentity` is injected, but still don't have required authentication. In that case, you will only have authenticated user when he has valid credentials (there are tests on RR / RC for these scenarios).  I think you can have application logic that only access identity in certain situations (e.g. you validated request and found out you need to load user specific profile). I can imagine it is not frequent scenario, but it's still valid to select authentication mechanism when auth/authZ is optional, WDYT?

Reply:
> As far as `Bearer`/`CodeFlow` is concerned, how would it work in a multi-tenant OIDC setup ?

@sberyozkin I believe `BearerTokenAuthorizationTest#testAnnotationBasedAuthMechSelection` shows just that. Please let me know if that's not what you meant.

Reply:
What I'm trying to say is that if we use `@Authenticated` we mean one of the authentication mechanism must be used for this method be allowed to be invoked. So when I say `@Bearer` I pretty much say exactly what `@Authentication` says except that a specific mechanism is requested, see what I mean or do I miss something ?

Reply:
> What I'm trying to say is that if we use `@Authenticated` we mean one of the authentication mechanisms must be used for this method be allowed to be invoked. 

Agreed. 

> So when I say `@Bearer` I pretty much say exactly what `@Authenticated` says except that a specific mechanism is requested, see what I mean or do I miss something ?

I mean, I agree, but I think we have same conclusion. It's valid case to have unauthenticated endpoint and access `SecurityIdentity`, therefore it is valid case to have endpoint annotated with `@CodeFlow` without `@Authenticated`. It can't be same. Maybe I need a little more from you, can you elaborate?

Reply:
What would `@CodeFlow without @Authenticated` mean then if on the method annotated with `@CodeFlow` an anonymous identity can be used ?

Reply:
> What would `@CodeFlow` without `@Authenticated` mean then if on the method annotated with `@CodeFlow` an anonymous identity can be used ?

Yeah, exactly (by the way, I'm testing scenarios when there is no RBAC but auth mechanism is selected via annotation)! The `@CodeFlow` enforces authentication mechanism, it means no other authentication mechanism can be used (which in my eyes makes sense - if you select auth mechanism, you don't want other one used... - and it's a same thing config approach does), so when you call `securityIdentity.getPrincipal().getName()`, anonymous identity is used and principal name is empty. But now imagine that you wont' annotate the endpoint with `@CodeFlow`, in that case if the request has basic authorization, basic auth mechanism is going to be used (or any other based on usual priorities and credentials type), and it will mean user could still authenticate.

I don't want to go wild, but I can imagine these user `admin` authenticated by basic auth is not a same one as `admin` authorized by code flow mechanism.

> if on the method annotated with @CodeFlow an anonymous identity can be used 

but the point is that it will be anonymous identity for `@CodeFlow`, but what if there is other credentials type for which it is not anonymous?

BTW I _think_ I remember we had Quarkus issues related to `SecurityIdentity` injection when no RBAC annotation was used, so it should show this is not just some unlikely scenario.

Reply:
@sberyozkin there is one more thing, more technical that I didn't want to mention as I hoped to convince you with above-mentioned arguments - currently you can't really combine security checks, so making it authenticated means no roles-allowed or denyall (in which case the identity will determine 401 or 403). however I can probably come with some trick to workaround that if necessary.

Reply:
I'm sorry Michal but I don't understand why in the presense of `quarkus-oidc`, `@Authenticated` and `@CodeFlow` mean different things

Reply:
> I'm sorry Michal but I don't understand why, in the presense of `quarkus-oidc`, `@Authenticated` and `@CodeFlow` mean different things

Let's consider application with basic, mTLS and code-flow auth mechanism and following resource

```
@Path("example")
public class ExampleResource {

    @Inject
    SecurityIdentity identity;

    @Path("fruits-1")
    @CodeFlow
    public List<String> fruits(@QueryParam("premium-stuff") boolean premiumStuff) {
        if (premiumStuff) {
            // only premium customers are allowed to access premium stuff
            if (identity.hasRole("premium-customer")) {
                return List.of("apple", "mandarin");
            } else {
                throw new ForbiddenException("Mandarins are only for premium customers");
            }
        } else {
            return List.of("apple");
        }
    }

    @Path("fruits-2")
    public List<String> fruitsWithoutCodeFlow(@QueryParam("premium-stuff") boolean premiumStuff) {
        if (premiumStuff) {
            // only premium customers are allowed to access premium stuff
            if (identity.hasRole("premium-customer")) {
                return List.of("apple", "mandarin");
            } else {
                throw new ForbiddenException("Mandarins are only for premium customers");
            }
        } else {
            return List.of("apple");
        }
    }

}
```

URL `/example/fruits-2?premium-stuff=true` can be accessed via basic/mTLS/code-flow while `/example/fruits-1?premium-stuff=true` won't ever be accessible by anything but code-flow. There are tests for it I believe.

Reply:
@michalvavrik Sure, so why can't `@CodeFlow` imply `@Authenticated` ?

To me, 
```
@Path("example")
public class ExampleResource {
    @Path("fruits-1")
    @GET
    @CodeFlow
    public List<String> fruits() {
    }
}
``` 

is identical to

```
@Path("example")
public class ExampleResource {
    @Path("fruits-1")
    @GET
    public List<String> fruits() {
    }
}
```
and
```
quarkus.http.auth.permission.code.paths=/example/fruits-1
quarkus.http.auth.permission.code.policy=authenticated
quarkus.http.auth.permission.code.auth-mechanism=code
```

Reply:
ah, now I understand what you mean @sberyozkin . Okay, we **can** make it same. In the end, it all goes down to this:

- in my origin example, path `/example/fruits-1` can be accessed without authentication and for `curl http://localhost:8080/example/fruits-1` you would get `apple`
- in your example, path `/example/fruits-1` can be accessed without authentication and for `curl http://localhost:8080/example/fruits-1` you would get 401

However this example is very simple, have you considered that RBAC annotations can also be used on CDI beans like here:

```
@ ApplicationScoped
public classs SecuredFruitsService extends FruitsService {

  @ Authenticated
  public List<String> fruits() {
     return List.of("apple", "mandarin");
  }

}

@ ApplicationScoped
public classs PublicFruitsService extends FruitsService {

  public List<String> fruits() {
     return List.of("apple");
  }

}
```

and what implementation of `FruitsService` can depend on application logic? Application flow can and cannot require `SecurityIdentity` depending on other factors. So by automatically requiring `@CodeFlow` endpoint authenticated you will decrease capabilities.

Reply:
>So by automatically requiring @CodeFlow endpoint authenticated you will decrease capabilities

But by saying `@CodeFlow` we say - this method must be authenticated by quarkus-oidc and its authorization code flow, the last example, the user can move `@CodeFlow` up to `SecuredFruitsService`

Reply:
> > So by automatically requiring @codeflow endpoint authenticated you will decrease capabilities
> 
> But by saying `@CodeFlow` we say - this method must be authenticated by quarkus-oidc and its authorization code flow, the last example, the user can move `@CodeFlow` up to `SecuredFruitsService`

I must admin I didn't realize that. Good point, but:

-  it means using CDI interceptors for you can't tell in `ServerRestHandler` if some CDI bean is going to be invoked. Considering Security Checks are performed in RESTEasy Reactive from `ServerRestHandler` and we have 2 (really 2!) sets of CDI interceptors - one is performing check and other in RR is making sure that the Security extension interceptors are skipped when check is performed by `ServerRestHandler`; allowing using `@CodeFlow` on other places than the REST endpoints means additional code / beans. Are you confident it is not edge case?
- **you authenticate once and once only per request**, which is reason why this feature do not work with proactive authentication (it is analogy, I don't connect your suggestion to proactive auth). Maybe you mentioned I added insurance that authentication mechanism is not selected **after** authentication has been attempted.
  -  If we put `@CodeFlow` on endpoint, situation is pretty straightforward, you can't authenticate in route handler you add manually with higher priority and that is it.
  - If you put `@CodeFlow` on some CDI bean, you must be pretty sure on no occasion (some JAX-RS filter, Hibernate validator, whatever invocation) for some reason didn't need the identity. It seems fragile to me. Annotating endpoints is safer.

Reply:
Hi @michalvavrik What I'm still not sure about is why you are concerned about `@CodeFlow` be seen as a concrete instance of `@Authenticated` and offer these variations. In my mind `@CodeFlow` `==` `@Authenticated` except that `@CodeFlow` insists on a specific authentication mechanism. Treating `@CodeFlow` (or `@Bearer` etc) and `@Authenticated` differently will only cause some ambiguities IMHO

Reply:
> Hi @michalvavrik What I'm still not sure about is why you are concerned about `@CodeFlow` be seen as a concrete instance of `@Authenticated` and offer these variations. In my mind `@CodeFlow` `==` `@Authenticated` except that `@CodeFlow` insists on a specific authentication mechanism. Treating `@CodeFlow` (or `@Bearer` etc) and `@Authenticated` differently will only cause some ambiguities IMHO

No @sberyozkin , this is not important for me at all, I just don't understand why would you like to enforce authentication when `@CodeFlow` is used, for I believe I provided concrete examples (please see `/example/fruits-1`) where you can select auth mechanism and but only require authentication if certain business logic is fulfilled and we will rob users of this capability.

I can't think of additional arguments I could provide, I do not insist on it, just raising arguments. I can settle on `@CodeFlow == @Authenticated` if you don't agree with my arguments.

Reply:
@michalvavrik As we've discussed offline, IMHO, the situation where unauthenticated requests will be able to enter `@CodeFlow` annotated methods and do whatever those methods do as anonymous identities if these methods do not happen to require an access to `SecurityIdentity` would cause the confusion and surprise IMHO.

We have HTTP Security Policy and path based authentication which will do fine when the proactive authentication is off and the security is required somewhere in the custom bean. Annotation based approach with `@CodeFlow` should not delay indefinitely the authentication IMHO - it may never happen as not every endpoint needs the identity access

Reply:
I'm still not convinced, but I don't want to go back. Let's work from now on with conclusion that @ CodeFlow == @ Authenticated by default.

> @michalvavrik As we've discussed offline, IMHO, the situation where unauthenticated requests will be able to enter @CodeFlow annotated methods and do whatever those methods do as anonymous identities if these methods do not happen to require an access to SecurityIdentity would cause the confusion and surprise IMHO.

Yes, but we still need to support combination for RBAC annotations (I am mainly talking about @RolesAllowed) with @ CodeFlow and considering RBAC annotations can't be combine, this is causing chicken-egg problem. It means we only need to add annotation transformer to method annotated with @ CodeFlow when no RBAC annotation is present, but RBAC annotations are also result of annotation transformation so ...

I'll work around this by introducing default `SecurityCheck` that is only applied when no other `SecurityCheck` is added for method. I'm not sure it's pretty, but I've been thinking about it for 3 days now and it's the only way.

> We have HTTP Security Policy and path based authentication which will do fine when the proactive authentication is off and the security is required somewhere in the custom bean. Annotation based approach with @CodeFlow should not delay indefinitely the authentication IMHO - it may never happen as not every endpoint needs the identity access

As said, I don't agree, but it doesn't matter, because you can always do

```
@CodeFlow
@PermitAll
public String myEndpoint() {
  return "unauthenticated";
}
```
and authentication will not be required. In case you are going to say we shouldn't allow that, please bear in mind that we simply need to support combination of `@CodeFlow` and `@RolesAllowed` (and `@PermissionsAllowed`) and it would be weird to add there exceptions like `@PermitAll` won't work.

Reply:
@sberyozkin ^^^

Reply:
@sberyozkin I really don't want to re-discuss it, we already agreed on solution, but I need to react on this one:

> Annotation based approach with @CodeFlow should not delay indefinitely the authentication

That's simply not precise. Please consider following HTTP Security Policy

```
quarkus.http.auth.permission."permissions".policy=permit
quarkus.http.auth.permission."permissions".paths=/my-endpoint
quarkus.http.auth.permission."permissions".auth-mechanism=bearer
```

Which I believe proves that current state of this PR is equal to what is possible and it is delaying nothing. It is simply selecting auth mechanism that is going to be used if and only if authentication is required.

Reply:
on the other hand, it wasn't that hard to implement, done. thanks for your input. I'm going to close this conversation as I believe that I addressed your concernes.